### PR TITLE
8335836: serviceability/jvmti/StartPhase/AllowedFunctions/AllowedFunctions.java fails with unexpected exit code: 112

### PR DESCRIPTION
**Pull request by sspitsyn**
The test has been fixed to:
 - add a guard against JVMTI_ERROR_WRONG_PHASE
 - replace `exit(err)` with `abort()` in the `check_jvmti_error()`
 
The JVMTI `VMDeath` event is enabled and a `RawMonitor` is introduced to serialize `ClassPrepare` and `VMDeath` events, and so, to prevent JVMTI_ERROR_WRONG_PHASE in the `ClassPrepare` events.

Testing:
 - run the test AllowedFunctions.java locally
 - TBD: submit a mach5 run for fixed test

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8335836](https://bugs.openjdk.org/browse/JDK-8335836): serviceability/jvmti/StartPhase/AllowedFunctions/AllowedFunctions.java fails with unexpected exit code: 112 (**Bug** - P4)

### Reviewers
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**)
 * [Alex Menkov](https://openjdk.org/census#amenkov) (@alexmenkov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20397/head:pull/20397` \
`$ git checkout pull/20397`

Update a local copy of the PR: \
`$ git checkout pull/20397` \
`$ git pull https://git.openjdk.org/jdk.git pull/20397/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20397`

View PR using the GUI difftool: \
`$ git pr show -t 20397`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20397.diff">https://git.openjdk.org/jdk/pull/20397.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20397#issuecomment-2259355847)

### CODE REVIEW

**Comment by sspitsyn** [comment]
Converting the native agent to C++ would simplify the test and allow to use utility functions from the `jvmti_common.hpp`.
But I wanted to make the WRONG_PHASE related fix to be clear for reviewers.
So, this conversion can be done as a separate step or as a part of a separate effort covered by an enhancement.

File: test/hotspot/jtreg/serviceability/jvmti/StartPhase/AllowedFunctions/libAllowedFunctions.c [hunk 1]
```diff
@@ -337,11 +352,20 @@ ClassPrepare(jvmtiEnv *jvmti, JNIEnv *env, jthread thread, jclass klass) {
     jvmtiPhase phase;
     intptr_t exp_val = 777;
     intptr_t act_val;
+    jvmtiError err;
+
+    // Block VMDeath event and other ClassPrepare events while this callback is executed.
+    // Sync with VMDeath event guards agains JVMTI_ERROR WRONG_PHASE.
+    err = (*jvmti)->RawMonitorEnter(jvmti, event_mon);
```
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/jvmti/StartPhase/AllowedFunctions/libAllowedFunctions.c:359]: I don't see how this fixes the problem. You can be exiting VMDeath when this monitor is entered. This does not block the get_thread_local() call below, which is what leads to the WRONG_PHASE. I think using a raw monitor is the right approach, but I was expecting to see VMDeath set a flag after grabbing the monitor that indicated VMDeath had been called. Then ClassPrepare should check this flag after grabbing the monitor and ignore the event if it is set.

File: test/hotspot/jtreg/serviceability/jvmti/StartPhase/AllowedFunctions/libAllowedFunctions.c [hunk 2]
```diff
@@ -337,11 +352,20 @@ ClassPrepare(jvmtiEnv *jvmti, JNIEnv *env, jthread thread, jclass klass) {
     jvmtiPhase phase;
     intptr_t exp_val = 777;
     intptr_t act_val;
+    jvmtiError err;
+
+    // Block VMDeath event and other ClassPrepare events while this callback is executed.
+    // Sync with VMDeath event guards agains JVMTI_ERROR WRONG_PHASE.
```
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/jvmti/StartPhase/AllowedFunctions/libAllowedFunctions.c:358]:
```suggestion
    // Sync with VMDeath event guards agains JVMTI_ERROR_WRONG_PHASE.
```

Comment (sspitsyn) [reply test/hotspot/jtreg/serviceability/jvmti/StartPhase/AllowedFunctions/libAllowedFunctions.c]: Fixed now, thanks!

File: test/hotspot/jtreg/serviceability/jvmti/StartPhase/AllowedFunctions/libAllowedFunctions.c (hunk 1 above)
Comment (sspitsyn) [reply test/hotspot/jtreg/serviceability/jvmti/StartPhase/AllowedFunctions/libAllowedFunctions.c]: Thank you for the comment.
Fixed the `get_thread_local()` issue and corrected phase check.
New flag is not really needed as the `phase == JVMTI_PHASE_DEAD` can play its role.
Please, let me know if you are not okay with that.

File: test/hotspot/jtreg/serviceability/jvmti/StartPhase/AllowedFunctions/libAllowedFunctions.c [hunk 3]
```diff
@@ -329,22 +330,47 @@ VMInit(jvmtiEnv *jvmti, JNIEnv* jnii, jthread thread) {
...
+    // Block VMDeath event and other ClassPrepare events while this callback is executed.
+    // Sync with VMDeath event guards against JVMTI_ERROR WRONG_PHASE.
+    err = (*jvmti)->RawMonitorEnter(jvmti, event_mon);
+    check_jvmti_error(jvmti, "ClassPrepare event: Failed in RawMonitorEnter", err);
 
     get_phase(jvmti, &phase);
     if (phase != JVMTI_PHASE_START && phase != JVMTI_PHASE_LIVE) {
-        printf("  ## Error: unexpected phase: %d, expected: %d or %d\n",
-               phase, JVMTI_PHASE_START, JVMTI_PHASE_LIVE);
+        if (phase != JVMTI_PHASE_DEAD) {
```
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/jvmti/StartPhase/AllowedFunctions/libAllowedFunctions.c:363]: After this check is made, what prevents JVMTI from moving to the JVMTI_PHASE_DEAD before the get_cur_thread() call below? I'm not sure if the phase is changed to JVMTI_PHASE_DEAD before or after calling VMDeath, but it seems in either case you can end up with the phase changing after making this check. There is nothing that stops the phase from changing while holding the raw monitor.

Comment (sspitsyn) [reply test/hotspot/jtreg/serviceability/jvmti/StartPhase/AllowedFunctions/libAllowedFunctions.c:363]: Thanks, Chris.
The `JvmtiExport::post_vm_death()` posts the `VMDeath` before changing the phase to `JVMTI_PHASE_DEAD`.
The `VMDeath` callback is blocked on the `RawMonitor` while it is grabbed by the `ClassPrepare` event callback.
But you are right, there is a sync gap in `JvmtiExport::post_vm_death()` between the `VMDeath` callback execution and phase changing. So, I'll go with your suggestion and introduce a flag.

Comment (sspitsyn) [reply test/hotspot/jtreg/serviceability/jvmti/StartPhase/AllowedFunctions/libAllowedFunctions.c]: Added `is_vm_dead` flag.

File: test/hotspot/jtreg/serviceability/jvmti/StartPhase/AllowedFunctions/libAllowedFunctions.c [hunk 4]
```diff
@@ -329,22 +331,52 @@ VMInit(jvmtiEnv *jvmti, JNIEnv* jnii, jthread thread) {
...
     intptr_t exp_val = 777;
     intptr_t act_val;
+    jvmtiError err;
+
+    // Block VMDeath event and other ClassPrepare events while this callback is executed.
+    // Sync with VMDeath event and check for is_vm_dead guard against JVMTI_ERROR WRONG_PHASE.
+    err = (*jvmti)->RawMonitorEnter(jvmti, event_mon);
+    check_jvmti_error(jvmti, "ClassPrepare event: Failed in RawMonitorEnter", err);
 
+    if (is_vm_dead == JNI_TRUE) {
```
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/jvmti/StartPhase/AllowedFunctions/libAllowedFunctions.c:363]:
```suggestion
    if (is_vm_dead) {
```

Comment (sspitsyn) [reply test/hotspot/jtreg/serviceability/jvmti/StartPhase/AllowedFunctions/libAllowedFunctions.c]: Thanks, updated now.

**Review by plummercj — APPROVED** [review]
Looks good.

**Comment by sspitsyn** [comment]
Thank you for review, Chris!

**Review by alexmenkov — APPROVED** [review]

**Comment by sspitsyn** [comment]
Thank you for review, Alex!

**Pull request closed without merging**